### ISSUE TITLE
Convolution and post operation of binary add or div very slow

### ISSUE THREAD

**Issue by edenBenhamou** [oneapi-src/oneDNN#1355]
Hi,

I'm checking the performance of a change I made in my convolution kernel. I'm using a forward convolution in oneDNN that is very efficient compared to my original code then I need to do a quantization manually. Previously it was made in cpp and now I'm trying to do it with post operations.

Here's part of my code, I have 4 post ops:
```
    memory::desc conv_src_md = memory::desc({conv_src_tz}, src_dt, tag::any);
    memory::desc conv_weights_md = memory::desc({conv_weights_tz}, dt::s8, tag::any);
    memory::desc conv_shifts_md = memory::desc({conv_shifts_tz}, dt::s32, tag::any);
    memory::desc conv_bias_md = memory::desc({conv_bias_tz}, dt::s32, tag::any);
    memory::desc conv_dst_md = memory::desc({conv_dst_tz}, dst_dt, tag::any);
    
    // create convolution descriptor
    convolution_forward::desc conv_desc = convolution_forward::desc(
            prop_kind::forward_inference,
            algorithm::convolution_direct,
            conv_src_md,
            conv_weights_md,
            conv_dst_md,
            conv_strides,
            conv_padding_l,
            conv_padding_r
    );

    //post op
    dnnl::post_ops conv_ops;

    //parameters for clamp
    float scale=1.0f;
    float alpha = (m_dst->getDataType()==DataType::I8) ? SCHAR_MIN : 0;
    float beta = (m_dst->getDataType()==DataType::I8) ? SCHAR_MAX : UCHAR_MAX;
    conv_ops.append_binary(algorithm::binary_div, conv_shifts_md);
    conv_ops.append_eltwise(scale, algorithm::eltwise_clip, alpha, beta);
    conv_ops.append_binary(algorithm::binary_add, conv_bias_md);
    conv_ops.append_eltwise(scale, algorithm::eltwise_clip, alpha, beta);
```

The performance of eltwise operation seems okay but the running time become much larger (-930% on small tensors, -12,000% on bigger tensors) when I'm adding the binary operation of div or add (I've tested each of them separately).
I was expecting that the operations fused to the convolution primitive in post ops would give me a better performance than my original quantization in cpp.
Can you explain me why it's happening? Why the binary post op are so much worse?
@igorsafo can you have a look? Is there anything I'm doing wrong?

I'm adding my verbosity output for more information:

Old version with quantization in cpp:

```
onednn_verbose,info,oneDNN v2.5.0 (commit def16767d9b1962d06ec8ad6f60e0b0a47dc58ea)
onednn_verbose,info,cpu,runtime:sequential
onednn_verbose,info,cpu,isa:Intel AVX-512 with Intel DL Boost
onednn_verbose,info,gpu,runtime:none
onednn_verbose,info,prim_template:operation,engine,primitive,implementation,prop_kind,memory_descriptors,attributes,auxiliary,problem_desc,exec_time
onednn_verbose,exec,cpu,convolution,jit_int8_1x1:avx512_core_vnni,forward_inference,src_s8::blocked:acdb:f0 wei_s8::blocked:ABcd4b16a4b:f3:s8m1 bia_undef::undef::f0 dst_s32::blocked:acdb:f0,attr-scratchpad:user ,alg:convolution_direct,mb1_ic128oc128_ih8oh8kh1sh1dh0ph0_iw8ow8kw1sw1dw0pw0,0.0258789
onednn_verbose,exec,cpu,convolution,jit_int8_1x1:avx512_core_vnni,forward_inference,src_s8::blocked:acdb:f0 wei_s8::blocked:ABcd4b16a4b:f3:s8m1 bia_undef::undef::f0 dst_s32::blocked:acdb:f0,attr-scratchpad:user ,alg:convolution_direct,mb1_ic128oc128_ih8oh8kh1sh1dh0ph0_iw8ow8kw1sw1dw0pw0,0.0109863
onednn_verbose,exec,cpu,convolution,jit_int8_1x1:avx512_core_vnni,forward_inference,src_s8::blocked:acdb:f0 wei_s8::blocked:ABcd4b16a4b:f3:s8m1 bia_undef::undef::f0 dst_s32::blocked:acdb:f0,attr-scratchpad:user ,alg:convolution_direct,mb1_ic128oc128_ih8oh8kh1sh1dh0ph0_iw8ow8kw1sw1dw0pw0,0.00683594
```

New version with quantization in post ops (slower performance):
```
onednn_verbose,info,oneDNN v2.5.0 (commit def16767d9b1962d06ec8ad6f60e0b0a47dc58ea)
onednn_verbose,info,cpu,runtime:sequential
onednn_verbose,info,cpu,isa:Intel AVX-512 with Intel DL Boost
onednn_verbose,info,gpu,runtime:none
onednn_verbose,info,prim_template:operation,engine,primitive,implementation,prop_kind,memory_descriptors,attributes,auxiliary,problem_desc,exec_time
onednn_verbose,create:cache_miss,cpu,convolution,ref_int8:any,forward_inference,src_s8::blocked:acdb:f0 wei_s8::blocked:abcd:f0 bia_undef::undef::f0 dst_s8::blocked:acdb:f0,attr-scratchpad:user attr-post-ops:binary_div:s32:2+eltwise_clip:-128:127+binary_add:s32:2+eltwise_clip:-128:127 ,alg:convolution_direct,mb1_ic128oc128_ih8oh8kh1sh1dh0ph0_iw8ow8kw1sw1dw0pw0,0.019043
onednn_verbose,exec,cpu,convolution,ref_int8:any,forward_inference,src_s8::blocked:acdb:f0 wei_s8::blocked:abcd:f0 bia_undef::undef::f0 dst_s8::blocked:acdb:f0,attr-scratchpad:user attr-post-ops:binary_div:s32:2+eltwise_clip:-128:127+binary_add:s32:2+eltwise_clip:-128:127 ,alg:convolution_direct,mb1_ic128oc128_ih8oh8kh1sh1dh0ph0_iw8ow8kw1sw1dw0pw0,5.31298
onednn_verbose,exec,cpu,convolution,ref_int8:any,forward_inference,src_s8::blocked:acdb:f0 wei_s8::blocked:abcd:f0 bia_undef::undef::f0 dst_s8::blocked:acdb:f0,attr-scratchpad:user attr-post-ops:binary_div:s32:2+eltwise_clip:-128:127+binary_add:s32:2+eltwise_clip:-128:127 ,alg:convolution_direct,mb1_ic128oc128_ih8oh8kh1sh1dh0ph0_iw8ow8kw1sw1dw0pw0,5.30908

```

**Comment by dzarukin** [comment]
Hi @edenBenhamou, thank you for the question.
The reason you're experiencing low performance is because it falls back to reference implementation (`ref_int8:any`), which is not an optimized kernel.
I've tried to reproduce the fall back on my side and I've noticed the following:
* If weights are passed with `tag::any`, everything works as expected and no fall back happens (see benchdnn command output 1 below).
* If weights are passed with not `tag::any` but with same tag as for optimized case, it will fall back to reference code. It happens because of [this logic](https://github.com/oneapi-src/oneDNN/blob/master/src/cpu/x64/jit_avx512_core_x8s8s32x_1x1_conv_kernel.cpp#L965) connected to extra flags. Only tag is not enough, extra should be defined as well. We don't expect users to define it on their side, thus, we expect int8 weights are passed with `tag::any` all the times. See benchdnn output 2.

Output 1:
```
run: --conv --allow-enum-tags-only=false --dir=FWD_I --cfg=s8s8s8 --attr-post-ops=div:s32:per_oc+clip:-128:127+add:s32:per_oc+clip:-128:127 --attr-scratchpad=user mb1ic128ih8oc128oh8kh1ph0
oneDNN implementation: jit_int8_1x1:avx512_core_vnni
onednn_verbose,info,oneDNN v2.5.0 (commit N/A)
onednn_verbose,info,cpu,runtime:sequential
onednn_verbose,info,cpu,isa:Intel AVX-512 with Intel DL Boost
onednn_verbose,info,gpu,runtime:none
onednn_verbose,info,prim_template:operation,engine,primitive,implementation,prop_kind,memory_descriptors,attributes,auxiliary,problem_desc,exec_time
oneDNN implementation: jit_int8_1x1:avx512_core_vnni
... (reorders) ...
onednn_verbose,exec,cpu,convolution,jit_int8_1x1:avx512_core_vnni,forward_inference,src_s8::blocked:acdb:f0 wei_s8::blocked:ABcd4b16a4b:f3:s8m1 bia_undef::undef::f0 dst_s8::blocked:acdb:f0,attr-scratchpad:user attr-post-ops:binary_div:s32:2+eltwise_clip:-128:127+binary_add:s32:2+eltwise_clip:-128:127 ,alg:convolution_direct,mb1_ic128oc128_ih8oh8kh1sh1dh0ph0_iw8ow8kw1sw1dw0pw0,0.0698242
... (reorders) ...
0:PASSED __REPRO: --conv --allow-enum-tags-only=false --dir=FWD_I --cfg=s8s8s8 --attr-post-ops=div:s32:per_oc+clip:-128:127+add:s32:per_oc+clip:-128:127 --attr-scratchpad=user mb1ic128ih8oc128oh8kh1ph0
```

Output 2:
```
run: --conv --allow-enum-tags-only=false --dir=FWD_I --cfg=s8s8s8 --stag=acdb --wtag=ABcd4b16a4b --dtag=acdb --attr-post-ops=div:s32:per_oc+clip:-128:127+add:s32:per_oc+clip:-128:127 --attr-scratchpad=user mb1ic128ih8oc128oh8kh1ph0
oneDNN implementation: ref_int8:any
onednn_verbose,info,oneDNN v2.5.0 (commit N/A)
onednn_verbose,info,cpu,runtime:sequential
onednn_verbose,info,cpu,isa:Intel AVX-512 with Intel DL Boost
onednn_verbose,info,gpu,runtime:none
onednn_verbose,info,prim_template:operation,engine,primitive,implementation,prop_kind,memory_descriptors,attributes,auxiliary,problem_desc,exec_time
oneDNN implementation: ref_int8:any
... (reorders) ...
onednn_verbose,exec,cpu,convolution,ref_int8:any,forward_inference,src_s8::blocked:acdb:f0 wei_s8::blocked:ABcd4b16a4b:f0 bia_undef::undef::f0 dst_s8::blocked:acdb:f0,attr-scratchpad:user attr-post-ops:binary_div:s32:2+eltwise_clip:-128:127+binary_add:s32:2+eltwise_clip:-128:127 ,alg:convolution_direct,mb1_ic128oc128_ih8oh8kh1sh1dh0ph0_iw8ow8kw1sw1dw0pw0,111.067
... (reorders) ...
0:PASSED __REPRO: --conv --allow-enum-tags-only=false --dir=FWD_I --cfg=s8s8s8 --stag=acdb --wtag=ABcd4b16a4b --dtag=acdb --attr-post-ops=div:s32:per_oc+clip:-128:127+add:s32:per_oc+clip:-128:127 --attr-scratchpad=user mb1ic128ih8oc128oh8kh1ph0
```

Notice that output 2 has `wtag` defined.
I can reproduce this behavior consistently on latest master and revision you are working on,

A question for you: could you double check that weights for a case with post-ops are created with `format_tag::any`, please?
If they seem to be created with `any`, could you share somehow the piece of code which creates convolution with post-ops with us, please? Thank you.

**Comment by dzarukin** [comment]
Hi @edenBenhamou, I think this is the sketchy place:
```
m_groupedData[idx].bias_memory= memory({conv_bias_tz, dt::s32, tag::nc}, eng, biasData);
```

Based on tag it seems that you are passing 2D tensor, and `conv_bias_tz` looks like `{MB, OC}`, where `MB` (based on verbose) is `1`.
The problem with the library that it doesn't have a uniform support for any kind of descriptor when passing it as a post-op. To trigger optimized version, your bias, as well as shifts, should be a 1D tensor. `conv_bias_tz` should look like `{OC}`. After that I would expect it start working fast as intended. Could you try it, please? Thank you.

**Comment by edenBenhamou** [comment]
Hi @dzarukin,
I've finally succeed to make it run on the optimized version and got a performance improvement.
The bias shouldn't be a 1D tensor, it gives me false results and didn't run the avx version neither. But, when I made it a 4D tensor {1,OC,1,1}, it worked perfectly. It seems that the broadcast couldn't work in optimized version without giving the same layout as the first source (convolution result) in Binary op.
Does it make sense to you?

Thanks for your help!
